### PR TITLE
COMMON: Expand cosine and sine table features

### PR DESCRIPTION
**Pull request by dafioram**
The cos/sin table class now has its array index operator
overridden so that it can be indexed rather than have
the internal table indexed directly.

This allows the checking and computing of the correct
indexes to be done internally.

The indexing in allows cos/sine of 0 to 2pi to be
obtained. Internally, cos/sine values of only 0 to pi/2
are computed.

An alternate way can have all 0 to 2pi cos/sine values
stored in memory, which will duplcate values, but require
less if checks for the correct index in operator[].

getByValue also allows the cosine/sine value
that is closests to the radian value returned.
Which is an approximation to cos and sine.

I am not sure why the current cosine Table only generates
cos values at angles where the cosine is positive.

If we feel good about the current cosineTable usage in dct, fft, and similar
and mine is too much of a deviation maybe we can have both classes and
rename the current one eosCosineTable or similar.

bladerunner and titanic both use a sine/cosine table so that is a good
case for a modified common cosine/sine table like this,

### CODE REVIEW

**Comment by dafioram** [comment]
Here is how my modified cosineTable for 4 bits does when evaluated at its 16 different points when its values is obtain by: the index, getValueByAngle, and the actual cosine function.

Index = 0 Radian Angle = 0 Table look up = 1 getValueByAngle = 1 Cosine = 1
Index = 1 Radian Angle = 0.392699 Table look up = 0.92388 getValueByAngle = 0.92388 Cosine = 0.92388
Index = 2 Radian Angle = 0.785398 Table look up = 0.707107 getValueByAngle = 0.707107 Cosine = 0.707107
Index = 3 Radian Angle = 1.1781 Table look up = 0.382683 getValueByAngle = 0.382683 Cosine = 0.382683
Index = 4 Radian Angle = 1.5708 Table look up = 0 getValueByAngle = 0 Cosine = 6.12323e-17
Index = 5 Radian Angle = 1.9635 Table look up = -0.382683 getValueByAngle = -0.382683 Cosine = -0.382683
Index = 6 Radian Angle = 2.35619 Table look up = -0.707107 getValueByAngle = -0.707107 Cosine = -0.707107
Index = 7 Radian Angle = 2.74889 Table look up = -0.92388 getValueByAngle = -0.92388 Cosine = -0.92388
Index = 8 Radian Angle = 3.14159 Table look up = -1 getValueByAngle = -1 Cosine = -1
Index = 9 Radian Angle = 3.53429 Table look up = -0.92388 getValueByAngle = -0.92388 Cosine = -0.92388
Index = 10 Radian Angle = 3.92699 Table look up = -0.707107 getValueByAngle = -0.707107 Cosine = -0.707107
Index = 11 Radian Angle = 4.31969 Table look up = -0.382683 getValueByAngle = -0.382683 Cosine = -0.382683
Index = 12 Radian Angle = 4.71239 Table look up = 0 getValueByAngle = 0 Cosine = -1.83697e-16
Index = 13 Radian Angle = 5.10509 Table look up = 0.382683 getValueByAngle = 0.382683 Cosine = 0.382683
Index = 14 Radian Angle = 5.49779 Table look up = 0.707107 getValueByAngle = 0.707107 Cosine = 0.707107
Index = 15 Radian Angle = 5.89049 Table look up = 0.92388 getValueByAngle = 0.92388 Cosine = 0.92388

**Comment by dafioram** [comment]
Same thing but for the sineTable:

Index = 0 Radian Angle = 0 Table look up = 0 getValueByAngle = 0 Sine = 0
Index = 1 Radian Angle = 0.392699 Table look up = 0.382683 getValueByAngle = 0.382683 Sine = 0.382683
Index = 2 Radian Angle = 0.785398 Table look up = 0.707107 getValueByAngle = 0.707107 Sine = 0.707107
Index = 3 Radian Angle = 1.1781 Table look up = 0.92388 getValueByAngle = 0.92388 Sine = 0.92388
Index = 4 Radian Angle = 1.5708 Table look up = 1 getValueByAngle = 1 Sine = 1
Index = 5 Radian Angle = 1.9635 Table look up = 0.92388 getValueByAngle = 0.92388 Sine = 0.92388
Index = 6 Radian Angle = 2.35619 Table look up = 0.707107 getValueByAngle = 0.707107 Sine = 0.707107
Index = 7 Radian Angle = 2.74889 Table look up = 0.382683 getValueByAngle = 0.382683 Sine = 0.382683
Index = 8 Radian Angle = 3.14159 Table look up = -0 getValueByAngle = -0 Sine = 1.22465e-16
Index = 9 Radian Angle = 3.53429 Table look up = -0.382683 getValueByAngle = -0.382683 Sine = -0.382683
Index = 10 Radian Angle = 3.92699 Table look up = -0.707107 getValueByAngle = -0.707107 Sine = -0.707107
Index = 11 Radian Angle = 4.31969 Table look up = -0.92388 getValueByAngle = -0.92388 Sine = -0.92388
Index = 12 Radian Angle = 4.71239 Table look up = 1 getValueByAngle = 1 Sine = -1
Index = 13 Radian Angle = 5.10509 Table look up = -0.92388 getValueByAngle = -0.92388 Sine = -0.92388
Index = 14 Radian Angle = 5.49779 Table look up = -0.707107 getValueByAngle = -0.707107 Sine = -0.707107
Index = 15 Radian Angle = 5.89049 Table look up = -0.382683 getValueByAngle = -0.382683 Sine = -0.382683

File: common/cosinetables.cpp [hunk 1]
```diff
@@ -20,29 +20,70 @@
  *
  */
 
-// Based on eos' cosine tables
-
 #include "common/cosinetables.h"
 #include "common/scummsys.h"
+#include <cmath>
```
Comment (sev-) [inline common/cosinetables.cpp]: We do not include STL headers.

File: common/cosinetables.cpp [hunk 2]
```diff
@@ -20,29 +20,70 @@
...
 	assert((bitPrecision >= 4) && (bitPrecision <= 16));
 
 	_bitPrecision = bitPrecision;
 
-	int m = 1 << _bitPrecision;
-	double freq = 2 * M_PI / m;
-	_table = new float[m / 2];
+	_nPoints = 1 << _bitPrecision;
+	_radResolution = 2.0 * M_PI / _nPoints;
+	_pi2Float = (float)2.0*M_PI;
```
Comment (sev-) [inline common/cosinetables.cpp]: Formatting is incorrect here.

File: common/cosinetables.cpp [hunk 3]
```diff
@@ -20,29 +20,70 @@
...
+	for (int i = 0; i < _size; i++)
+		_table[i] = cos(i * _radResolution);
+
+}
+
+float CosineTable::operator[] (int index) const { 
+	if (index < _size)
+		// [0,pi/2)
+		return _table[index];
+	if (index > _size && index < 2*_size)
```
Comment (sev-) [inline common/cosinetables.cpp]: Formatting is not correct, and in the lines below.

File: common/sinetables.cpp [hunk 4]
```diff
@@ -20,29 +20,70 @@
...
 	assert((bitPrecision >= 4) && (bitPrecision <= 16));
 
 	_bitPrecision = bitPrecision;
 
-	int m = 1 << _bitPrecision;
-	double freq = 2 * M_PI / m;
-	_table = new float[m / 2];
+	_nPoints = 1 << _bitPrecision;
+	_radResolution = 2.0 * M_PI / _nPoints;
+	_pi2Float = (float)2.0*M_PI;
```
Comment (sev-) [inline common/sinetables.cpp]: Incorrect formatting.

File: common/sinetables.cpp [hunk 5]
```diff
@@ -20,29 +20,70 @@
  *
  */
 
-// Based on eos' sine tables
-
 #include "common/scummsys.h"
 #include "common/sinetables.h"
+#include <cmath>
```
Comment (sev-) [inline common/sinetables.cpp]: Another STL header. Should be replaced.

**Comment by dafioram** [comment]
Fixed spacings and removed stl.

Added another method getValueByAngleDeg() which is just the degree version instead of the radian version (getValueByAngle is the radian version).

The current CosineTable and SineTable defined the indexes differently then I have so when comparing them we only get the same values for the first 2^bits / 4 indexes (we define the resolution the same).

**Comment by dafioram** [comment]
I did some time testing. I tried to evaluate 0.8125*2*pi by the several methods to see how long it would take. I call it over and over 1 million times using floats. The sineTable used 16 bit precision (the largest):

sine call: 28ms (the base line)
operator[] call: 7 ms
getValueByAngle() call: 32 ms
Direct Array access: 4.5ms

So the extra checking the operator[] call to figure out the indexes cost a bit compared to the direct access, but it was still way better than the sine call and the getValueByAngle was terrible. Maybe the rounding call is making that function take so long, either way the sine function is preferable for speed over getValueByAngle.

**Comment by dafioram** [comment]
I have changed the commit so that:
1. Dropped getValueByAngle and by Degree functions since they are slower than cos/sine call
2. getTable() now returns the exact same table values as it did before this commit so that this commit is purely an enhancement.

**Comment by dafioram** [comment]
Changed operator[] to at() because cosTable->operator[](x) doesn't look as good as cosTable->at(x) for Common::CosineTable *cosTable.

File: common/cosinetables.cpp [hunk 6]
```diff
@@ -24,25 +24,45 @@
 
 #include "common/cosinetables.h"
 #include "common/scummsys.h"
+#include <math.h>
```
Comment (bonki) [inline common/cosinetables.cpp]: Why was this added?

**Comment by dafioram** [comment]
Removed math.h include.

**Comment by bonki** [comment]
Thanks! @sev-, OK to merge?

**Comment by sev-** [comment]
Yes, merging now. Thanks!
